[mlir][linalg] Enable fuse consumer

cxy-1993 opened the pull request:
This patch adds support for consumer fusion to the tiling interface, and implements fuse consumers on FuseIntoContainingOp.

- Add interface method 'getIterDomainTilePositionFromOperandPosition' to tiling interface which get iteration domain position from operand position.
- Add interface method 'getTiledImplementationFromOperandPosition' to tiling interface which generate tiled implementation according to operand position.
- Implemented the above two methods and supported consumer fusion for FuseIntoContainingOp.
- Add tests for fuse consumer interface

cxy-1993 committed:
[mlir][linalg] Enable fuse consumer

This patch adds support for consumer fusion to the tiling interface, and
implements fuse consumers on FuseIntoContainingOp.

- Add interface method 'getIterDomainTilePositionFromOperandPosition' to
tiling interface which get iteration domain position from operand
position.
- Add interface method 'getTiledImplementationFromOperandPosition' to
tiling interface which generate tiled implementation according to
operand position.
- Implemented the above two methods and supported consumer fusion for
FuseIntoContainingOp.

cxy-1993 commented:
This PR is a re-submission from (https://github.com/llvm/llvm-project/pull/85528), with changes to include additional tests.  Please help me review this patch. @MaheshRavishankar @ftynse @Abhishek-Varma @nicolasvasilache

banach-space commented:
For those of us less familiar with the fusion logic, could you please provide a high level example with "BEFORE" and "AFTER"? You can use pseudo IR for that (I can use the implementation to reason about the finer details).

nujaa commented:
Could the behaviour of   `-test-linalg-fuse-consumer` be reproduced in transform dialect ? E.g.  with TileUsingForallOp + FuseOp ?

MaheshRavishankar commented:
Thanks @cxy-1993 for the change.  I have comments, but just want to state at the outset that these arent a reaction to this PR or the effort to follow through on recommendations, but just unfortunate evolution of things.

What is being done here is what I was trying to avoid. To test the interface methods now there is an implementation of tile and fuse in a test folder. I don't see how this is useful. 
https://github.com/llvm/llvm-project/pull/88712 is the place to test such a thing. This is why I didnt ask for tests on the original PR. So even if this is landed this way, all this is doing is adding more things to delete after the fact. I dont have the heart to block this change, but I really dont think this is a good outcome here.

ftynse commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+
+    // Check consumer don't use more than one result of containingOp.
+    Value bridge(nullptr);
+    SmallVector<unsigned> operandNums;
```
This vector only ever has one element, consider replacing with a simple value. Also see below about `OpOperand &`.

ftynse commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    // Check consumer doon't use scf.forall's output as init.
+    SmallVector<Value> dpsInits = llvm::to_vector<4>(
```
Nit: `llvm::map_to_vector`.

ftynse replied:
Also, I don't see the need to materialize a vector here. `getDpsInits()` returns a range that should be suitable for `llvm::is_contained`. If not, that range is implicitly convertible to `ValueRange` that is suitable.

ftynse commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    // Try to get all containing op result's position from iter domain position.
+    llvm::SmallVector<std::pair<llvm::SmallVector<OpFoldResult>,
+                                llvm::SmallVector<OpFoldResult>>>
```
No need to prefix with `llvm::`.

ftynse commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+      return;
+    }
+
+    auto tiledOps = tileAndFuseResult->tiledOps;
```
Expand auto here and below.

ftynse commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+          bridge = opd;
+        } else if (bridge != opd) {
+          emitError(consumerOp->getLoc())
+              << "consumer's operand use more than one containingOp's result";
```
```suggestion
              << "consumer's operands use more than one containingOp's result";
```

ftynse commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+          newforallOp->getResult(forallOp.getOutputs().size() +
+                                 consumerResult.index()));
+    }
+    forallOp.erase();
```
This must go through the rewriter.

ftynse reviewed (changes requested):
Thanks for iterating on this!

It appears that some code was copy-pasted from #88712. At least, this is my impression after leaving multiple comments for identical issues in the code. If it is the case, please make sure all those issues are also addressed. You are also most welcome to collaborate with the author of that PR on a single branch and send one, unified PR.

ftynse commented:
> What is being done here is what I was trying to avoid. To test the interface methods now there is an implementation of tile and fuse in a test folder. I don't see how this is useful.

I agree that this isn't exactly useful, because it introduces a lot of additional logic beyond what actually needs to be tested. Moreover, the newly added tests only test this behavior indirectly, and may end up rather hard to debug should the behavior regress.

There is a much simpler testing strategy: implement a test pass that calls the newly added interface methods on all operations that implement the interface and outputs the results of the call as diagnostic remarks (attributes can be printed, values can be identified by their locations). This is straightforward and fits into a couple dozen lines of code. We could also consider a unit test. It was rather obvious to me, and I probably should have elaborated on that...

ftynse commented:
> So even if this is landed this way, all this is doing is adding more things to delete after the fact.

I don't think this code is going to be deleted, but rather moved. So not much waste outside of reviewer time.

cxy-1993 commented on mlir/include/mlir/Interfaces/TilingInterface.td:
```diff
@@ -82,15 +82,34 @@ def TilingInterface : OpInterface<"TilingInterface"> {
           used to obtain the tiled implementation of the operation.
         }],
-        /*retType=*/"LogicalResult",
+        /*retType=*/"::mlir::LogicalResult",
```
DONE

cxy-1993 commented on mlir/test/Dialect/Linalg/test-fuse-consumer.mlir:
```diff
@@ -0,0 +1,103 @@
+#map1 = affine_map<(d0)[s0] -> (d0 * s0)>
+#map2 = affine_map<(d0)[s0] -> (-(d0 * s0) + 64, s0)>
+// CHECK-LABEL: func.func @fuse_tileable_consumer
+// CHECK-SAME: %[[CHUNK_SIZE:[0-9a-z]+]]: index
```
I tried your method, but there's an issue with matching because filecheck will match extra characters due to having the same data type on the same line, causing errors.

ftynse replied:
I see, thanks for trying. Maybe let's update to `[0-9A-Za-z_]+` then, which should capture all supported names.

cxy-1993 commented on mlir/test/lib/Dialect/Linalg/TestLinalgFuseConsumer.cpp:
```diff
@@ -0,0 +1,294 @@
+    });
+
+    if (!consumerOp || !containingOp || walkRes.wasInterrupted()) {
+      emitError(getOperation()->getLoc())
```
Good idea, done.

cxy-1993 commented on mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:
```diff
@@ -132,14 +132,66 @@ struct LinalgOpTilingInterface
+    if (!indexingMap.isPermutation()) {
+      SmallVector<Range> iterationDomain =
+          tilingInterfaceOp.getIterationDomain(b);
+      for (const auto &&[index, value] : llvm::enumerate(iterationDomain)) {
```
"const" has been removed. Thanks for sharing knowledge about &&. Here, I want to use references to prevent reconstruction, and I don't care whether it's a lvalue reference or rvalue reference. It looks like a lvalue reference is used here.

cxy-1993 commented:
> For those of us less familiar with the fusion logic, could you please provide a high level example with "BEFORE" and "AFTER"? You can use pseudo IR for that (I can use the implementation to reason about the finer details).

Thank you for your suggestion. A complete transform/pass submission should indeed include relevant descriptions. However, as discussed above, the implementation here is a test that is about to be moved or deleted. I feel it's unnecessary to provide excessive description of inputs and outputs here. We can refine that in this submission(https://github.com/llvm/llvm-project/pull/88712).

cxy-1993 commented:
> Could the behaviour of `-test-linalg-fuse-consumer` be reproduced in transform dialect ? E.g. with TileUsingForallOp + FuseOp ?

Yes. Considering it's about to be moved or deleted, I think this can be addressed in the next submission（https://github.com/llvm/llvm-project/pull/88712）.

Abhishek-Varma reviewed (commented):
Just adding this here as a proposal to consider if everyone thinks this would indeed be a better way forward :-

I believe having just the infra of `TilingInterface.td` + `TilingInterfaceImpl.cpp` (basically the reverted changes) as the separate base commit in the in-flight PR [88712](https://github.com/llvm/llvm-project/pull/88712) would be the most preferable way to go.

Reasons for the same :-
1. The PR I added indeed makes use of the Interface methods to achieve the fusion for both scf.for and scf.forall.
2. The same PR also adds the necessary tests based on a test transform dialect op.
3. Most of the review comments by @ftynse  has already been addressed - so @cxy-1993 can avoid having to redo the same set of things here.

After [88712](https://github.com/llvm/llvm-project/pull/88712) lands the only set of actionable item left would be to define an official transform dialect that does the same job as the test transform dialect I've already added in my PR - something perhaps @cxy-1993 can take up thereafter.

This avoids any unnecessary code movements and makes it cleaner.

cxy-1993 commented:
> Just adding this here as a proposal to consider if everyone thinks this would indeed be a better way forward :-
> 
> I believe having just the infra of `TilingInterface.td` + `TilingInterfaceImpl.cpp` (basically the reverted changes) as the separate base commit in the in-flight PR [88712](https://github.com/llvm/llvm-project/pull/88712) would be the most preferable way to go.
> 
> Reasons for the same :-
> 
> 1. The PR I added indeed makes use of the Interface methods to achieve the fusion for both scf.for and scf.forall.
> 2. The same PR also adds the necessary tests based on a test transform dialect op.
> 3. Most of the review comments by @ftynse  has already been addressed - so @cxy-1993 can avoid having to redo the same set of things here.
> 
> After [88712](https://github.com/llvm/llvm-project/pull/88712) lands the only set of actionable item left would be to define an official transform dialect that does the same job as the test transform dialect I've already added in my PR - something perhaps @cxy-1993 can take up thereafter.
> 
> This avoids any unnecessary code movements and makes it cleaner.

The proposal is also fine with me. Since @MaheshRavishankar @ftynse all agree  on this, I'll close this PR. Please cherry pick the interface part to your PR @Abhishek-Varma , thanks!

BTW, Thanks again to @ftynse  for reviewing this patch.
